snappy/faststreams.nim

```diff
@@ -0,0 +1,147 @@
+import
+  std/strutils,
+  pkg/faststreams/[inputs, multisync, outputs],
+  "."/[codec, encoder, exceptions],
+  ../snappy
+
+export
+  inputs, multisync, outputs, codec, exceptions
+
+{.push raises: [Defect].}
+
+proc checkCrcAndAppend(
+    output: OutputStream, data: openArray[byte], crc: uint32): bool {.
+    raises: [Defect, IOError].}=
+  if maskedCrc(data) == crc:
+    output.write(data)
+    return true
+
+proc compress*(input: InputStream, output: OutputStream) {.
+    raises: [Defect, InputTooLarge, IOError].} =
+  ## Compress all bytes of `input`, writing into `output` and flushing at the end.
+  ##
+  ## Input length must not exceed `maxUncompressedLen == 2^32-1` or
+  ## `InputTooLarge` will be raised. Other errors are raised as they happen on
+  ## the given streams.
+  doAssert input.len.isSome, "TODO: support actual .. streams"
+  let
+    lenU32 = checkInputLen(input.len.get).valueOr:
+      raiseInputTooLarge()
+    maxCompressed = maxCompressedLen(input.len.get).valueOr:
+      raiseInputTooLarge()
+
+  # TODO https://github.com/status-im/nim-faststreams/issues/28
+  # output.ensureRunway maxCompressed
+  output.write lenU32.toBytes(Leb128).toOpenArray()
+
+  var
+    # TODO instead of a temporary buffer, use `getWriteableBytes` once it
```

Comment:
The rest of the implementation is already full of `unsafeAddr` calls, so what is your concerns about using `getWritableBytes`?

Reply:
it gave an immutable `openArray` as discussed - it's marked TODO so as to get a stable core in place before choosing the right workaround